src/ExpressionEvaluator/Core/Source/ExpressionCompiler/CustomTypeInfo.cs

```diff
@@ -1,37 +1,196 @@
+        // and {tupleNames} is a UTF8 encoded string of the names separated by '|'.
+        internal static ReadOnlyCollection<byte> Encode(
+            ReadOnlyCollection<byte> dynamicFlags,
+            ReadOnlyCollection<string> tupleElementNames)
+        {
+            if ((dynamicFlags == null) && (tupleElementNames == null))
+            {
+                return null;
+            }
+            var builder = ArrayBuilder<byte>.GetInstance();
+            if (dynamicFlags == null)
+            {
+                builder.Add(0);
+            }
+            else
+            {
+                int length = dynamicFlags.Count;
+                if (length > byte.MaxValue)
+                {
+                    // TODO: Test.
+                    throw new NotImplementedException();
+                }
+                builder.Add((byte)length);
+                builder.AddRange(dynamicFlags);
+            }
+            if (tupleElementNames != null)
+            {
+                var bytes = EncodeNames(tupleElementNames);
+                builder.AddRange(bytes);
+            }
+            return new ReadOnlyCollection<byte>(builder.ToArrayAndFree());
+        }
+
+        internal static void Decode(
+            Guid payloadTypeId,
+            ReadOnlyCollection<byte> payload,
+            out ReadOnlyCollection<byte> dynamicFlags,
+            out ReadOnlyCollection<string> tupleElementNames)
+        {
+            dynamicFlags = null;
+            tupleElementNames = null;
+
+            if ((payload == null) || (payloadTypeId != PayloadTypeId))
+            {
+                return;
+            }
+
+            int length = payload[0];
+            if (length > 0)
+            {
+                dynamicFlags = new ReadOnlyCollection<byte>(CopyBytes(payload, 1, length));
+            }
+
+            int start = length + 1;
+            if (start < payload.Count)
+            {
+                tupleElementNames = DecodeNames(payload, start);
+            }
+        }
+
+        private const char NameSeparator = '|';
+
+        private static ReadOnlyCollection<byte> EncodeNames(ReadOnlyCollection<string> names)
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var builder = pooledBuilder.Builder;
+            foreach (var name in names)
+            {
+                if (builder.Length > 0)
+                {
+                    builder.Append(NameSeparator);
+                }
+                if (name != null)
+                {
+                    builder.Append(name);
+                }
+            }
+            var str = pooledBuilder.ToStringAndFree();
+            return new ReadOnlyCollection<byte>(Encoding.UTF8.GetBytes(str));
+        }
 
-        public CustomTypeInfo(Guid payloadTypeId, ReadOnlyCollection<byte> payload)
+        private static ReadOnlyCollection<string> DecodeNames(ReadOnlyCollection<byte> bytes, int start)
```

Comment:
I don't know if the string encoding includes a preceding length. I didn't include a length for the string since that would have meant dealing with byte ordering for strings longer than 256 bytes. If we need to add different metadata in the future, we can change the `Guid` and the change the format then.